File: backend/src/main/java/com/uznai/service/QuizSessionService.java
package com.uznai.service;

import com.uznai.dto.request.StartQuizSessionRequest;
import com.uznai.dto.request.SubmitAnswerRequest;
import com.uznai.dto.response.QuizQuestionResponse;
import com.uznai.dto.response.QuizSessionResponse;
import com.uznai.security.UserPrincipal;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

public interface QuizSessionService {
    QuizSessionResponse startSession(StartQuizSessionRequest request, UserPrincipal userPrincipal);
    
    QuizSessionResponse getSession(UUID sessionId, UserPrincipal userPrincipal);
    
    Page<QuizQuestionResponse> getSessionQuestions(UUID sessionId, UserPrincipal userPrincipal, Pageable pageable);
    
    void submitAnswer(UUID sessionId, UUID questionId, SubmitAnswerRequest request, UserPrincipal userPrincipal);
    
    QuizSessionResponse completeSession(UUID sessionId, UserPrincipal userPrincipal);
    
    void expireSession(UUID sessionId, UserPrincipal userPrincipal);
    
    Page<QuizSessionResponse> getUserSessions(UserPrincipal userPrincipal, Pageable pageable);
} 

File: frontend/src/services/quiz-session-service.ts
import { ApiClient } from '@/lib/api/client';
import type {
    QuizSession,
    Question,
    StartQuizSessionRequest,
    SubmitAnswerRequest,
} from '@/types/quiz-session';

export const quizSessionService = {
    startSession: async (quizId: string): Promise<QuizSession> => {
        return ApiClient.post<QuizSession>(`/api/v1/quizzes/${quizId}/sessions`, {
            quizId,
        });
    },

    getSession: async (sessionId: string): Promise<QuizSession> => {
        return ApiClient.get<QuizSession>(`/api/v1/sessions/${sessionId}`);
    },

    getSessionQuestions: async (sessionId: string, page = 0, size = 10): Promise<{
        content: Question[];
        totalPages: number;
        totalElements: number;
    }> => {
        return ApiClient.get(`/api/v1/sessions/${sessionId}/questions?page=${page}&size=${size}`);
    },

    submitAnswer: async (
        sessionId: string,
        questionId: string,
        request: SubmitAnswerRequest
    ): Promise<void> => {
        await ApiClient.post(
            `/api/v1/sessions/${sessionId}/questions/${questionId}/answers`,
            request
        );
    },

    completeSession: async (sessionId: string): Promise<QuizSession> => {
        return ApiClient.post<QuizSession>(`/api/v1/sessions/${sessionId}/complete`, {});
    },

    expireSession: async (sessionId: string): Promise<void> => {
        await ApiClient.post(`/api/v1/sessions/${sessionId}/expire`, {});
    },

    getUserSessions: async (page = 0, size = 10): Promise<{
        content: QuizSession[];
        totalPages: number;
        totalElements: number;
    }> => {
        return ApiClient.get(`/api/v1/users/me/sessions?page=${page}&size=${size}`);
    },
}; 